Add KPP Langmuir enhancement factor using WAVEWATCHIII Stokes drift

sbrus89 opened the pull request:
Up to 6 wavenumber-partitioned surface Stokes drift fields are coupled to MPAS-O and used to reconstruct the Stokes drift profile and compute surface-layer averages. The surface layer average of the Stokes drift is used to compute the Langmuir number and KPP enhancement factor from Van Roekel et al. (2012).

(This is a duplicate of #5, which was mistakenly marked as closed after I push the wrong branch to OceanNGD/add-ww3)

sbrus89 committed:
Add partitioned Stokes drift information to ww3 namelists

sbrus89 committed:
Add partitioned stokes drift variable coupling

sbrus89 committed:
Fix compile bugs in wav_comp_mct.F90 and ww3_cpl_indices.f90

sbrus89 committed:
Add significant wave height as coupled field

sbrus89 committed:
Change wave coupling variable names

 - get rid of wave prefix

sbrus89 committed:
Remove original surface stokes drift coupling fields

sbrus89 committed:
Remove outdated comments

sbrus89 committed:
Add mpaso namelist and streams options for active wave langmuir

sbrus89 commented:
@vanroekel, I'd appreciate you looking this one over pretty carefully when you get a chance.

sbrus89 commented:
The relevant MPAS-O changes are here:

https://github.com/sbrus89/MPAS-Model/compare/6af0a6e55f4c9b97d098d6fedda5b46182299caa..2f10b1d0af03a7d08e2cd95f1926a274b5fde2a6

vanroekel commented:
@sbrus89 thanks for the PR already.  Happy to see this happen!  I’m looking into this now and have a few high level requests/questions

1) Could you post the reference you follow to compute Us?  Is it the Brevik et al paper?
2) Could you plot a few sample stokes drift profiles and the resulting diffusivity?
3) why only allow partitions of 3,4,6 and not 5?
4) I would advocate that we keep the computation of stokes drift wholly separate from KPP.  All of the options about configurations the calculation I would not name them with reference to KPP

I’ll keep digging in.

vanroekel commented on components/mpas-ocean/bld/build-namelist:
```diff
@@ -644,11 +644,14 @@ add_default($nl, 'config_cvmix_kpp_nonlocal_with_implicit_mix');
 if ($ocn_wave eq 'true' ) {
 	add_default($nl, 'config_cvmix_kpp_use_theory_wave', 'val'=>'.false.');
 	add_default($nl, 'config_cvmix_kpp_use_active_wave', 'val'=>'.true.');
+	add_default($nl, 'config_cvmix_kpp_n_wavenumber_partitions');
```
I would suggest separating the wavenumber partition from KPP

vanroekel commented on components/ww3/src/cpl/ww3_cpl_indices.f90:
```diff
@@ -18,9 +18,20 @@ module ww3_cpl_indices
-  integer :: index_w2x_Sw_ustokes
-  integer :: index_w2x_Sw_vstokes
-  integer :: index_w2x_Sw_hstokes
+  integer :: index_w2x_Sw_ustokes_wavenumber_1
```
mostly a curiosity that could have been anywhere, what does `Sw` stand for throughout the code?

philipwjones replied:
In the coupler or coupled context, the convention is that S is for a state variable (F for flux) and the small letter is the component (w for wave). And stuff that happens in the coupler is x, so w2x is wave to coupler.

vanroekel replied:
thanks @philipwjones good to know!

sbrus89 commented:
Thanks for the comments @vanroekel.

>     1. Could you post the reference you follow to compute Us?  Is it the Brevik et al paper?

What I implemented here isn't the Brevik et al. approach. I ended going with the wavenumber-partiioned approach described in [Li et al. 2019](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2019MS001810I). This uses the surface Stokes drift evaluated at different wavenumbers, which can then be used to reconstruct the full profile (see Eqn. 9):
![Screen Shot 2021-06-08 at 2 28 45 PM](https://user-images.githubusercontent.com/3514023/121245671-ea1db680-c865-11eb-839c-de3398ae1bed.png)


>     2. Could you plot a few sample stokes drift profiles and the resulting diffusivity?

I'm still debugging right now, but I will do this once things are fully working.

>     3. why only allow partitions of 3,4,6 and not 5?

I reached out to Brandon Reichl and he graciously shared the wavenumbers he has been using. The values he gave me were for 3, 4, and 6.

>     4. I would advocate that we keep the computation of stokes drift wholly separate from KPP.  All of the options about configurations the calculation I would not name them with reference to KPP

Good suggestion; I agree. I will make this change.

sbrus89 committed:
Add significant wave height coupling variable to ww3 namelist

sbrus89 committed:
Broadcast partitioned Stokes drift information

 - Prevents error with setting gridded output and coupling variable
   flags

sbrus89 committed:
Zero out significant wave height at non-sea points

sbrus89 committed:
Update WW3 submodule

 - Rebase RTD fixes

sbrus89 committed:
Remove USP as an output variable for ww3

sbrus89 committed:
Subtree merge of Stokes drift MPAS-Ocean changes

sbrus89 committed:
Update cime submodule

 - Fix MCT build error that occurs with newer version of cime

sbrus89 committed:
Updates for G case + waves with Stokes drift coupling

 - Change PE layout

 - Add Langmuir number to ocean streams

 - Turn off wave model output to avoid MPI errors

 - Uncomment KPP enhancement factor calculations

 - Modify compset to include active sea ice

sbrus89 committed:
Update WW3 submodule

 - Extend the length of mesh filename variable

sbrus89 committed:
Fixes log file redirection error

 - Caused by the hardcoded unit number in w3grid

 - Also updates WW3 submodule to include w3grid calling argument
   change (unit numbers added)

sbrus89 committed:
Add ice damping limits and relative wind namelist parameters

sbrus89 committed:
Add UOST inputs to namelist and switch files

sbrus89 committed:
Add e3sm_wav_developer test group

sbrus89 committed:
Add unit number to read_stations_file subroutine

sbrus89 committed:
Use IC2/IS2 switches in WW3

sbrus89 committed:
Reorganize Langmuir number calculations

sbrus89 committed:
Add boundaryLayerDepth to restart stream

sbrus89 committed:
Add peak direction and frequency to coupling variables

sbrus89 committed:
Add B-case with waves compset

sbrus89 committed:
Add stokes drift omp private variables in mpas_ocn_vmix_cvmix.F

sbrus89 committed:
Fix typo for peak frequency/direction import
